refactor: benchmarking process

hoeck opened the pull request:
Finally I got down to it :sweat_smile: 

This resolves #497 by:

* adding support for multiple benchmarks
* splitting the current benchmark into 4 separate benchmarks (parse/assert x loose/strict)
* adding a separate test for each benchmark to ensure correct api usage
* adding an interactive gh-pages hosted graph to view/filter/sort benchmark results [preview here](https://hoeck.github.io/typescript-runtime-type-benchmarks/)

Please have a look at it @moltar

~~I have not resolved all linting issues (not sure how up-to-date the rules are).~~

The nodejs version range is 17, 16, 14, 12 - the gh actions should changed to reflect that so that results for all those nodejs versions are created. I was not sure whether node 10 is still needed.

I removed the static links to the csv, json and svg results from the readme. The results are still generated though. Any idea whether to keep this or not?

Also I removed the outlier benchmark split, as this can be easily achieved by deselecting the "Loose Assertion" benchmark.

Before or after merging you need to enable Github pages support from within the `docs/` folder of your repository: https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site

moltar commented:
OMG @hoeck you are a legend! 

---

>  I was not sure whether node 10 is still needed.

Yeah, probably can be dropped at this point.

> I removed the static links to the csv, json and svg results from the readme. The results are still generated though. Any idea whether to keep this or not?

I think maybe we can embed a screenshot of the results, which links to the page with the results. Just to make it more obvious? I kind-a liked embedded images in the README, but having an interactive one is MUCH better, ofc.

I just wanted to have an immediate visual stimulation when a repo is loaded 🗡️ 

> Also I removed the outlier benchmark split, as this can be easily achieved by deselecting the "Loose Assertion" benchmark.

Perfect!

> Before or after merging you need to enable Github pages support

Done! ✅ 

---

I noticed that the generated files now have a Node version that is very specific. Prior I used the `14.x` pattern, but now it includes the full Node version. Will this post a problem with versioning?

moltar commented on cases/bueno.ts:
```diff
@@ -15,16 +25,90 @@ const dataType = object({
+addCase(
+  'bueno',
+  'parseSafe',
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
What do you think of using a shared type, e.g.:

```
type UnknownData = any
```

And then just import it everywhere. This will give us a better name and context for why it's `any`, and also avoid the `eslint-disable-next-line` eye sores.

hoeck replied:
Oh of course that makes absolute sense. Will fix that

hoeck commented:
> OMG @hoeck you are a legend!
Thanks :grin: 

> I think maybe we can embed a screenshot of the results, which links to the page with the results. Just to make it more obvious? I kind-a liked embedded images in the README, but having an interactive one is MUCH better, ofc.
> 
> I just wanted to have an immediate visual stimulation when a repo is loaded dagger
Completely agree. Was toying with the same idea, getting visitors interested and clicking the link too but was unsure and wanted to get the PR finished :grin: .
I'll see if I can create the same graph again and add it as an image link to the readme.

> > Also I removed the outlier benchmark split, as this can be easily achieved by deselecting the "Loose Assertion" benchmark.
> 
> Perfect!
> 
> > Before or after merging you need to enable Github pages support
> 
> Done! white_check_mark
> 
> I noticed that the generated files now have a Node version that is very specific. Prior I used the `14.x` pattern, but now it includes the full Node version. Will this post a problem with versioning?

Yes that will definitely result in a lot of files in the results folder as the node minor versions go up.

Please note that I have kept files at `/results` just for backward compatibility. They're not used in the interactive report and are not included in the readme any more.

The `app.tsx` is fetching separate accumulated benchmark results from `docs/results/node_<major_version_number>.json`.

So, should I keep the svg/json/csv stuff at `results` (fixing the filenames to not include the node minor version) or remove it completely?

moltar commented:
> Please note that I have kept files at /results just for backward compatibility.

Meh, let's nuke them. Git history will preserve them, if necessary. Otherwise, they will just confuse others.

> So, should I keep the svg/json/csv stuff at results (fixing the filenames to not include the node minor version) or remove it completely?

Lets just nuke everything that is no longer relevant!

hoeck commented:
Added a "top 3 of each benchmark and median for the rest" preview graph and put that on top of the readme. You can see it here: https://github.com/hoeck/typescript-runtime-type-benchmarks/tree/refactor-benchmarks

What do you think? Experimented with lots Vega visualizations (pie charts, several compressed bar charts) but found this to be the most useful & pretty one.

> Lets just nuke everything that is no longer relevant!

Removed everything. Also cleaned up the benchmark generation process. Turns out there is no need for me to read and combine the benchmark files - I can just use the benchmark results directly.

Finally I added some small fixes to the interactive viewer to also work well in dark mode (water.css already supports this, only Vega needs some hand holding).

Now I only need to fix the build errors :facepalm:

moltar commented:
Woohoo!! 😁
